examples/resource-forall.cpp

```diff
@@ -0,0 +1,376 @@
+//----------------------------------------------------------------------------//
+{
+  std::cout << "\n Running RAJA GPU vector addition on 2 seperate streams...\n";
+#if defined(RAJA_ENABLE_CUDA)
+  RAJA::resources::Cuda res_gpu1;
+  RAJA::resources::Cuda res_gpu2;
+  using EXEC_POLICY = RAJA::cuda_exec_async<GPU_BLOCK_SIZE>;
+#elif defined(RAJA_ENABLE_HIP)
+  RAJA::resources::Hip res_gpu1;
+  RAJA::resources::Hip res_gpu2;
+  using EXEC_POLICY = RAJA::hip_exec_async<GPU_BLOCK_SIZE>;
+#endif
+
+  int* d_a1 = res_gpu1.allocate<int>(N);
+  int* d_b1 = res_gpu1.allocate<int>(N);
+  int* d_c1 = res_gpu1.allocate<int>(N);
+
+  int* d_a2 = res_gpu2.allocate<int>(N);
+  int* d_b2 = res_gpu2.allocate<int>(N);
+  int* d_c2 = res_gpu2.allocate<int>(N);
+
+  res_gpu1.memcpy(d_a1, a, sizeof(int)* N);
+  res_gpu1.memcpy(d_b1, b, sizeof(int)* N);
+
+  res_gpu2.memcpy(d_a2, a, sizeof(int)* N);
+  res_gpu2.memcpy(d_b2, b, sizeof(int)* N);
+
+
+  RAJA::forall<EXEC_POLICY>(res_gpu1, RAJA::RangeSegment(0, N), 
+    [=] RAJA_DEVICE (int i) { 
+    d_c1[i] = d_a1[i] + d_b1[i]; 
+  });    
+
+  RAJA::forall<EXEC_POLICY>(res_gpu2, RAJA::RangeSegment(0, N), 
+    [=] RAJA_DEVICE (int i) { 
+    d_c2[i] = d_a2[i] + d_b2[i]; 
+  }); 
+
+  res_gpu1.memcpy(c, d_c1, sizeof(int)*N );
+
+  res_gpu2.memcpy(c_, d_c2, sizeof(int)*N );
+
+  checkResult(c, N);
+  checkResult(c_, N);
+
+  res_gpu1.deallocate(d_a1);
+  res_gpu1.deallocate(d_b1);
+  res_gpu1.deallocate(d_c1);
+
+  res_gpu2.deallocate(d_a2);
+  res_gpu2.deallocate(d_b2);
+  res_gpu2.deallocate(d_c2);
+}
+
+
+//----------------------------------------------------------------------------//
+// RAJA::cuda/hip_exec policy with waiting event.... 
+//----------------------------------------------------------------------------//
+{
+  std::cout << "\n Running RAJA GPU vector with dependency between two seperate streams...\n";
+#if defined(RAJA_ENABLE_CUDA)
+  // _raja_res_defres_start
+  RAJA::resources::Cuda res_gpu1;
+  RAJA::resources::Cuda res_gpu2;
+  RAJA::resources::Host res_host;
+
+  using EXEC_POLICY = RAJA::cuda_exec_async<GPU_BLOCK_SIZE>;
+  // _raja_res_defres_end
+#elif defined(RAJA_ENABLE_HIP)
+  RAJA::resources::Hip res_gpu1;
+  RAJA::resources::Hip res_gpu2;
+  RAJA::resources::Host res_host;
+
+  using EXEC_POLICY = RAJA::hip_exec_async<GPU_BLOCK_SIZE>;
+#endif
+
+  // _raja_res_alloc_start
+  int* d_array1 = res_gpu1.allocate<int>(N);
+  int* d_array2 = res_gpu2.allocate<int>(N);
+  int* h_array  = res_host.allocate<int>(N);
+  // _raja_res_alloc_end
+
+  // _raja_res_k1_start
+  RAJA::forall<EXEC_POLICY>(res_gpu1, RAJA::RangeSegment(0,N),
+    [=] RAJA_HOST_DEVICE (int i) {
+      d_array1[i] = i;
+    }
+  );
+  // _raja_res_k1_end
+
+  // _raja_res_k2_start
+  RAJA::resources::Event e = RAJA::forall<EXEC_POLICY>(res_gpu2, RAJA::RangeSegment(0,N),
+    [=] RAJA_HOST_DEVICE (int i) {
+      d_array2[i] = -1;
+    }
+  );
+  // _raja_res_k2_end
+
+  // _raja_res_wait_start
+  res_gpu2.wait_for(&e);
```

Comment:
I know this is late feedback (and likely more appropriate for CAMP) but is there a reason why we pass a pointer rather than a reference? We can always convert the reference to a pointer as necessary. I don't think pointers are used in a public interface anywhere else. 

https://github.com/LLNL/camp/blob/1668c8995e956570d7b4f39ac0c832f5ebb11170/include/camp/resource.hpp#L90
```cpp
void wait_for(Event& e) { m_value->wait_for(&e); }
```

And perhaps the same for EventProxy?

https://github.com/LLNL/camp/blob/1668c8995e956570d7b4f39ac0c832f5ebb11170/include/camp/resource.hpp#L165

```cpp
EventProxy(Res& r) :
  resource_{&r}
{}
```

Reply:
Not that I can think of, in fact my brain completely tripped over this question and did a "whaaaa? don't we already do that?" loop for a moment.  I'm getting a feeling we are going to want to tweak a few things in this interface, or more likely extend then eventually deprecate etc.  This should definitely go on the list. Would you mind throwing this over to camp as an issue?